smtppipeline.go

```diff
@@ -274,6 +255,116 @@ func matchStepFromCfg(node config.Node) (SMTPPipelineStep, error) {
 	return res, nil
 }
 
+type destinationCond struct {
+	value  string
+	regexp *regexp.Regexp
+}
+
+func (cond destinationCond) matches(addr string) bool {
+	if cond.regexp != nil {
+		return cond.regexp.MatchString(addr)
+	}
+
+	if strings.Contains(cond.value, "@") {
+		return cond.value == addr
+	}
+
+	parts := strings.Split(addr, "@")
+	if len(parts) != 2 {
+		return false
+	}
+
+	// Domains are always case-sensetive.
+	return strings.ToLower(parts[1]) == strings.ToLower(cond.value)
+}
+
+type destinationStep struct {
+	conds    []destinationCond
+	substeps []SMTPPipelineStep
+}
+
+func (step *destinationStep) Pass(ctx *module.DeliveryContext, msg io.Reader) (io.Reader, bool, error) {
+	ctxCpy := *ctx
```

Comment:
We should probably deep-copy all fields here. This would prevent filters from inside `destination` to mutate the context for consumers outside (e.g. if a filter replaces a recipient).

Reply:
Oh, right. Totally forgot that there are reference-based fields other than `To`.